### PR TITLE
Land Smooth Game Action

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Keep the train rolling.

### CODE REVIEW

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+        if (result->Error != GA_ERROR::OK)
+        {
+            return result;
+        }
+
+        if (isExecuting)
+        {
+            audio_play_sound_at_location(SOUND_PLACE_ITEM, _coords.x, _coords.y, centreZ);
+        }
+        res->Cost += res->Cost;
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: This doesn't seem right.

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Could move this up to where MakeResult is.

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+
+        auto normRange = _range.Normalise();
+        // Cap bounds to map
+        auto l = std::max(normRange.GetLeft(), 32);
+        auto t = std::max(normRange.GetTop(), 32);
+        auto r = std::clamp(normRange.GetRight(), 0, (MAXIMUM_MAP_SIZE_TECHNICAL - 1) * 32);
+        auto b = std::clamp(normRange.GetBottom(), 0, (MAXIMUM_MAP_SIZE_TECHNICAL - 1) * 32);
+        auto validRange = MapRange{ l, t, r, b };
+
+        // Play sound (only once)
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Comment seems out of place.

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+        return SmoothLand(false);
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        return SmoothLand(true);
+    }
+
+private:
+    GameActionResult::Ptr QueryExecute(bool isExecuting) const
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Unused member

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+#include "../world/Sprite.h"
+#include "../world/Surface.h"
+#include "GameAction.h"
+
+DEFINE_GAME_ACTION(LandSmoothAction, GAME_COMMAND_EDIT_LAND_SMOOTH, GameActionResult)
+{
+private:
+    CoordsXY _coords;
+    MapRange _range;
+    uint8_t _selectionType;
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Uninitialized member

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+    bool _isLowering;
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Uninitialized member

File: src/openrct2/actions/LandSmoothAction.hpp [hunk 7]
```diff
@@ -0,0 +1,649 @@
...
+                z = map_get_corner_height(newBaseZ, newSlope, 0);
+                res->Cost += SmoothLandRowByCorner(isExecuting, { validRange.GetLeft(), validRange.GetTop() }, z, 32, 32, 2, 0);
+                z = map_get_corner_height(newBaseZ, newSlope, 3);
+                res->Cost += SmoothLandRowByCorner(
+                    isExecuting, { validRange.GetLeft(), validRange.GetTop() }, z, -32, 32, 1, 3);
+                z = map_get_corner_height(newBaseZ, newSlope, 1);
+                res->Cost += SmoothLandRowByCorner(
+                    isExecuting, { validRange.GetLeft(), validRange.GetTop() }, z, 32, -32, 3, 1);
+                break;
+            }
```
Comment (ZehMatt) [inline src/openrct2/actions/LandSmoothAction.hpp]: Missing default case for invalid selection type.

File: src/openrct2/actions/LandSmoothAction.hpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2/actions/LandSmoothAction.hpp]: Yeah it should be result

**Pull request merged**